### PR TITLE
Add acme_dns_server recipe

### PR DESCRIPTION
**Pull request by andrewda**
- [x] [Consider API security](https://github.com/osuosl-cookbooks/osl-acme/pull/17#discussion_r633996515)
- [x] [Consider failure handling](https://github.com/osuosl-cookbooks/osl-acme/pull/17#issuecomment-850036989)
- [ ] Add multi-node tests (separate PR)
  - [ ] Test getting same certificates across multiple nodes

## `records` databag:

```
{
  "id": "records",
  "records": [
    {
      "domain": "test1.example.org"
    },
    {
      "domain": "test2.example.org",
      "alt_names": [
        "test2-san1.example.org",
        "test2-san2.example.org"
      ]
    }
  ]
}
```

## `credentials` databag:

```
{
  "id": "credentials",
  "credentials": {
    "test1.example.org": {
      "subdomain": "c8d6aeae-3f21-4786-b243-98bbd7c526a5",
      "username": "cbfed1bb-c0b9-4b24-b212-5b95caa38f98",
       "key": "Kq3cjUDjfngndZm6ChkFZv3FY97e37vxYbfNbVXy"
     },
    "test2.example.org": {
      "subdomain": "2acca63e-1c34-4860-95f3-ba208dd8b0bc",
      "username": "c6853765-5036-4f01-9325-c7b97ee0fb2e",
      "key": "GJ62cjdfMfzkZpYhJCmpbdwMaBywmfg7EVJN8Umm"
    },
    "test2-san1.example.org": {
      "subdomain": "773aacab-9539-4d59-b5e6-76718261e078",
      "username": "0a59ceba-4648-4b91-b3d0-879160ab5bdf",
      "key": "sHXgTTZtxKXhnK23XusVRZUbGpvae43cZnjNSn52"
    },
    "test2-san2.example.org": {
      "subdomain": "d70eafbf-fa48-41fd-9496-bb6fa973c535",
      "username": "e3cf6541-81ee-46a6-a2f0-6a2ddd85871e",
      "key": "V6fJJe6WKz2zhvAV3VyNuWNFUpa8zWUTpcrmyzW5"
    }
  }
}
```

### CODE REVIEW

**Comment by ramereth** [comment]
I'm getting the following error when starting named on a fresh build:

```
/etc/named.conf:22: open: /etc/crypto-policies/back-ends/bind.config: file not found
```

**Comment by andrewda** [comment]
@ramereth Is that on CentOS 7 or 8? I've been testing almost exclusively on 8 for now, so maybe the bind config on 7 should be different?

**Comment by ramereth** [comment]
Ah, I was testing on 7 and haven't tried on 8. I noticed you're not using the osl-dns cookbook for setting up bind. That might help the situation and would be needed anyway. Although that cookbook is more setup for our DNS servers so it might need some adjustment.

**Comment by andrewda** [comment]
That's a good point. The `osl-dns` cookbook will need to be refactored a bit because it's currently cloning the `zonefiles` repo in the master recipe. Would it make sense to make a new recipe, like `generic_master` that would be responsible for getting things setup without any of the specific zonefiles?

**Comment by ramereth** [comment]
@andrewda it might be better to just make a specific `acme` recipe which does what you need for this on the bind server side.

**Comment by andrewda** [comment]
This will only be in the test cookbook, though, where we're setting up a local mock DNS setup. Unless I'm mistaken, the prod DNS would be configured through the zonefiles repository like normal. If you'd still like that on `osl-dns` that would work, but my thought process was that there would be some generic DNS configuration that applies to the master primary DNS server (`util.osuosl.org` IIRC) as well as for the ACME tests here.

**Comment by ramereth** [comment]
@andrewda yeah we should have a `default` recipe which does the basics for setting up a generic server. Please go ahead and do that and then use it here. At some point I need to see how this will be implemented on util for a production server. Ideally we should have a recipe just for that.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+host    all             all             0.0.0.0/0               trust
+  EOF
+  notifies :reload, 'service[postgresql]', :immediately
+end
+
+#
+# Configure acme-dns
+#
+
+execute 'add acme-dns ip' do
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:50]: I recommend you use the new `osl_fakenic` resource for this from the base cookbook.

**Comment by andrewda** [comment]
Another thing to consider: currently if any alt name fails the challenge, the entire certificate will fail. Would it be better to simply request another certificate without the failing alt names so a minor DNS change (malicious or accidental) doesn't mess up a bigger certificate?

**Comment by ramereth** [comment]
> Another thing to consider: currently if any alt name fails the challenge, the entire certificate will fail. Would it be better to simply request another certificate without the failing alt names so a minor DNS change (malicious or accidental) doesn't mess up a bigger certificate?

What does the current failure look like when that happens?

File: libraries/acme.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+def create_client(private_key, directory, contact)
+  require 'acme-client'
+  client = Acme::Client.new(private_key: private_key, directory: directory)
+  client.new_account(contact: contact, terms_of_service_agreed: true)
+  client
+end
+
+def request_record(acme_dns_api)
+  require 'acme-client'
```
[the remark this answers is not in the document]
Comment (ramereth) [reply libraries/acme.rb:23]: I think if you tried converting this into a proper Cookbooks helper module (like we've done in other cookbooks), that would be the first step and needs to happen as well. Can you please make this change before I do a big review on this?

**Comment by andrewda** [comment]
> What does the current failure look like when that happens?

It basically just complains that not all the challenges passed:

```
================================================================================                                                                                 
Error executing action `create` on resource 'get_acme_cert[test2.example.org]'                                                                                   
================================================================================                                                                                 
                                                                                                                                                                            
Acme::Client::Error                                                                                                                                              
-------------------                                                                                                                                              
Order's status ("invalid") was not ready 
```

When the challenges pass successfully, the order's status is "ready" and it can be fulfilled.

File: recipes/acme_dns_server.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  system true
+end
+
+directory '/etc/acme-dns' do
+  recursive true
+  action :create
+end
+
+acme_dns_version = node['osl-acme']['acme-dns']['version']
+remote_file '/etc/acme-dns/acme-dns.tar.gz' do
```
Comment (ramereth) [inline recipes/acme_dns_server.rb:31]: This file should be downloaded to this location, *not* in `/etc`.
```suggestion
remote_file "#{Chef::Config[:file_cache_path]}/acme-dns.tar.gz" do
```

File: recipes/acme_dns_server.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  source "https://github.com/joohoi/acme-dns/releases/download/v#{acme_dns_version}/acme-dns_#{acme_dns_version}_linux_386.tar.gz"
```
Comment (ramereth) [inline recipes/acme_dns_server.rb:32]: You should be using the `amd64` 64bit release, not the 32bit.

File: recipes/acme_dns_server.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+db_config = data_bag_item('osl_acme', 'database')
+
+template '/etc/acme-dns/config.cfg' do
+  source 'config.cfg.erb'
+  variables(dns_interface: dns_address,
+            pg_host: db_config['host'],
+            pg_user: db_config['user'],
+            pg_pass: db_config['pass'],
+            pg_dbname: db_config['dbname'])
+  notifies :restart, 'systemd_unit[acme-dns.service]'
```
Comment (ramereth) [inline recipes/acme_dns_server.rb:61]: This template should have a more restrictive file permission and also be set to `sensitive`

File: templates/default/config.cfg.erb [hunk 6]
```diff
@@ -0,0 +1,64 @@
+[general]
+# DNS interface. Note that systemd-resolved may reserve port 53 on 127.0.0.53
```
Comment (ramereth) [inline templates/default/config.cfg.erb:2]: Can we remove all comments and trim this config down a bit? Also add the managed by Chef header.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+cookbook_file '/etc/named.conf'
+
+directory '/var/named/master' do
+  recursive true
+end
+
+template '/var/named/master/db.example.org' do
+  source 'db.example.org.erb'
+  notifies :restart, 'bind_service[default]', :immediately
+end
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:29]: This should be a new recipe in `osl-dns` Can you please do that?

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+  only_if <<-EOF
+    sleep 15 &&
+    psql -U #{db_config['user']} #{db_config['dbname']} -c "
+      SELECT COUNT(*) FROM records
+      WHERE #{test_subdomains}" |
+    tr -d '\n' |
+    grep -P 'count -------     0'
+  EOF
+end
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:93]: Why is this needed?

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+postgresql_user db_config['user'] do
+  database db_config['dbname']
+  password db_config['pass']
+  action :create
+end
+
+file '/var/lib/pgsql/12/data/pg_hba.conf' do
+  content <<-EOF
+local   all             all                                     trust
+host    all             all             0.0.0.0/0               trust
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:50]: This is not the proper way to do this.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  EOF
+  notifies :reload, 'service[postgresql]', :immediately
+end
+
+#
+# Configure acme-dns
+#
+
+osl_fakenic 'lo' do
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:59]: Please do create a fake nic on `lo`. Just create a new `eth1 interface for testing.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+#
+
+records = data_bag_item('osl_acme', 'records')['records']
+records.each do |record|
+  get_acme_cert record['domain'] do
+    contact 'mailto:andrewda@osuosl.org'
+    acme_directory node['acme']['dir']
+    acme_dns_api node['osl-acme']['acme-dns']['api']
+    cert_path "/tmp/#{record['domain']}.pem"
+    key_path "/tmp/#{record['domain']}.key"
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:106]: This should not be where the certs are located.

File: resources/get_acme_cert.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
+require 'net/http'
+require 'openssl'
+require 'uri'
+require 'json'
+
+resource_name :get_acme_cert
+provides :get_acme_cert
```
Comment (ramereth) [inline resources/get_acme_cert.rb:7]: IMO this resource should be called `dns_acme_cert`

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 13]
```diff
@@ -0,0 +1,110 @@
+node.default['osl-acme']['pebble']['always_valid'] = false
+node.default['osl-acme']['pebble']['command'] = '/usr/local/bin/pebble -config /opt/pebble/test/config/pebble-config.json -dnsserver :8053'
+
+build_essential 'osl-acme-test' do
+  compile_time true
+end
+
+chef_gem 'bcrypt' do
+  compile_time true
+end
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:10]: Where is this being used?

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  get_acme_cert record['domain'] do
+    contact 'mailto:andrewda@osuosl.org'
+    acme_directory node['acme']['dir']
+    acme_dns_api node['osl-acme']['acme-dns']['api']
+    cert_path "/tmp/#{record['domain']}.pem"
+    key_path "/tmp/#{record['domain']}.key"
+
+    alt_names record['alt_names']
+  end
+end
```
Comment (ramereth) [inline test/cookbooks/acme_test/recipes/acme_dns_client.rb:110]: I feel like this should be in a resource so it can be reused easily on other hosts.

**Review by ramereth — COMMENTED** [review]
A great start!

I'm still having issues with this not working on CentOS 7 btw.

File: recipes/acme_dns_server.rb [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+systemd_unit 'acme-dns.service' do
+  content <<-EOF.gsub(/^\s+/, '')
```
[the remark this answers is not in the document]
Comment (detjensrobert) [reply recipes/acme_dns_server.rb:65]: (for reference: the squiggly heredoc was introduced in Ruby 2.3 and dedents the contents -- docs on it are [here](https://docs.ruby-lang.org/en/master/doc/syntax/literals_rdoc.html#label-Here+Documents+-28heredocs-29).)

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 9 above)
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:50]: Heh, I figured. The `postgresql_access` resource wasn't working so I did this as a temporary hack until I figure that out.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 13 above)
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:10]: I'm using it [here](https://github.com/osuosl-cookbooks/osl-acme/pull/17/files#diff-d7a45c50a60c1a3fe0b4b6cdb65835078a8eaed41c66cc75e7ae60090731f4ac) and [here](https://github.com/osuosl-cookbooks/osl-acme/pull/17/files#diff-0322745e4ad572bf464535011935b09f891c0f934d6461b521bf3dc9b9843191R70) (tests only) to generate the password hashes dynamically and store them in the database. We could alternatively store the password hashes in the databag, but I figured it was better to keep the databag as similar to production as possible (and would make updating the tests more difficult).

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 8 above)
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:93]: Without this, authentication won't work. We need to add the records to the database.

If you're asking why we're doing this instead of generating new usernames/keys for every test, we could do it that way, but it's less similar to how this will work in production. In production we'll have the subdomains, usernames, keys stored in the databag after being generated by someone at OSL. This simulates someone having already created the keys so they can be referenced later from the databag.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 14 above)
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:110]: Agreed. How should we handle the `cert_path` and `key_path` properties? Should we accept a `cert_dir` and `key_dir` where all the *.pem and *.key files are saved?

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 7 above)
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:29]: So currently this is just entirely test data and only used in testing, I'm a bit confused why it needs to go in `osl-dns`. In production, we'll just add `_acme-challenge` records to the zonefile, right? Where else would the recipe be used?

File: recipes/acme_dns_server.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewda) [reply recipes/acme_dns_server.rb]: @ramereth do you have an example of a helper method for `systemd_unit` definitions? I looked around our repos and couldn't find anything.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 8 above)
Comment (ramereth) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb]: OK, I see what's going on here. Is there any reason why you can't just make an sql dump of the state you want and then just do an import instead of doing it this way? Take a look at how we did this in the proj-snowdrift cookbook (specifically the blog). I think that would be cleaner than doing this way.

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb (hunk 13 above)
Comment (ramereth) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:10]: If you made a sql dump like I suggested earlier, would this no longer be needed?

**Comment by ramereth** [comment]
@andrewda can you please update and resolve the conflicts?

File: libraries/helpers.rb [hunk 16]
```diff
@@ -0,0 +1,94 @@
+module OslAcme
+  module Cookbook
+    module Helpers
+      require 'net/http'
+      require 'openssl'
+      # require 'acme-client'
```
Comment (ramereth) [inline libraries/helpers.rb:6]: Let's remove this comment if it's not needed here.

File: libraries/helpers.rb [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+      require 'uri'
+      require 'json'
+
```
Comment (ramereth) [inline libraries/helpers.rb:9]: Can you include additional comments throughout this file please?

File: resources/dns_acme_cert.rb [hunk 18]
```diff
@@ -0,0 +1,71 @@
+require 'openssl'
```
Comment (ramereth) [inline resources/dns_acme_cert.rb:1]: This should not be included here and should be done in an `action_class` block instead I think.

File: resources/dns_acme_cert.rb [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  private_key = get_key_or_nil(new_resource.key_path)
+
+  # If key does not exist, generate one and save to file
+  unless private_key
+    private_key = OpenSSL::PKey::RSA.new(2048)
+
+    file new_resource.key_path do
+      content private_key.export()
```
Comment (ramereth) [inline resources/dns_acme_cert.rb:37]: Shouldn't this be `sensitive true` ?

File: resources/dns_acme_cert.rb [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+property :acme_dns_api, String, default: lazy { node['osl-acme']['acme-dns']['api'] }
+property :cert_path, String
+property :key_path, String
+
+action :create do
+  existing_cert = get_cert_or_nil(new_resource.cert_path)
+  identifiers = [new_resource.domain, *new_resource.alt_names]
+
+  # Check if an existing certificate exists that expires in more than 24 hours
+  if existing_cert
```
Comment (ramereth) [inline resources/dns_acme_cert.rb:22]: Can you please include some additional comments throughout this file?

File: resources/dns_acme_cert.rb [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    valid = perform_challenge(
+      authorization,
+      domain_creds['subdomain'],
+      domain_creds['username'],
+      domain_creds['key'],
+      new_resource.acme_dns_api
+    )
+
+    unless valid
+      puts 'Challenge failed!'
```
Comment (ramereth) [inline resources/dns_acme_cert.rb:60]: I feel like this should be using `raise` here if something did fail instead of doing it this way.

File: spec/unit/recipes/acme_dns_server_spec.rb [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+            'test2-san1.example.org': {
+              'subdomain': '773aacab-9539-4d59-b5e6-76718261e078',
+              'username': '0a59ceba-4648-4b91-b3d0-879160ab5bdf',
+              'key': 'sHXgTTZtxKXhnK23XusVRZUbGpvae43cZnjNSn52',
+            },
+          }
+        )
+        stub_data_bag_item('osl_acme', 'database').and_return(id: 'database', host: '127.0.0.1', user: 'testuser', pass: 'testpass', dbname: 'testdb')
+
+        stub_command('iptables --list | grep "POST /register"').and_return(false)
```
Comment (ramereth) [inline spec/unit/recipes/acme_dns_server_spec.rb:47]: I don't think this is needed anymore.

**Review by ramereth — CHANGES_REQUESTED** [review]

File: attributes/default.rb [hunk 23]
```diff
@@ -1,18 +1,14 @@
...
-  Environment=PEBBLE_VA_ALWAYS_VALID=1
-  Environment=PEBBLE_VA_NOSLEEP=1
-  Environment=PEBBLE_WFE_NONCEREJECT=0
-  ExecStart=/usr/local/bin/pebble -config /opt/pebble/test/config/pebble-config.json
+default['osl-acme']['acme-dns']['version'] = '0.8'
+default['osl-acme']['acme-dns']['checksum'] = 'f5c031a78ea867a40c3b7cdb1d370f423bdf923e79a9607e44dabdc4dbda6a05'
+default['osl-acme']['acme-dns']['api'] = 'http://192.168.10.1'
+default['osl-acme']['acme-dns']['domain'] = 'acme-dns.example.org'
+default['osl-acme']['acme-dns']['nsname'] = 'ns.acme-dns.example.org'
+default['osl-acme']['acme-dns']['nsadmin'] = 'test.example.org'
```
[the remark this answers is not in the document]
Comment (andrewda) [reply attributes/default.rb:12]: Agreed, good point

File: test/cookbooks/acme_test/recipes/acme_dns_client.rb [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+include_recipe 'osl-acme::default'
+
+#
+# Get certificates for the specified hosts
+#
+
+dns_acme_certs 'Get ACME certificates' do
+  records data_bag_item('osl_acme', 'records')['records']
+
+  contact 'mailto:andrewda@osuosl.org'
```
[the remark this answers is not in the document]
Comment (andrewda) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:75]: In order to make it compatible with the existing `acme.contact` attribute, we should use the same format:

https://github.com/osuosl-cookbooks/osl-acme/blob/1d34676f315acdf50698692efc9a59ca24969f4d/recipes/default.rb#L20

[the remark this answers is not in the document]
Comment (ramereth) [reply test/cookbooks/acme_test/recipes/acme_dns_client.rb:75]: For this test recipe, we should just use the default for now though.

**Comment by ramereth** [comment]
I get the following error on a fresh run on CentOS 7:
```
       Recipe: acme_test::acme_dns_client
         * dns_acme_certs[Get ACME certificates] action create
           * dns_acme_cert[test1.example.org] action create
             * file[/etc/pki/tls/test1.example.org.key] action create
        - create new file /etc/pki/tls/test1.example.org.key
        - update content in file /etc/pki/tls/test1.example.org.key from none to c01b34
        - suppressed sensitive resource
        - restore selinux security context
             
             ================================================================================
             Error executing action `create` on resource 'dns_acme_cert[test1.example.org]'
             ================================================================================
             
             RuntimeError
             ------------
             Challenge failed for test1.example.org
             
             Cookbook Trace: (most recent call first)
             ----------------------------------------
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_cert.rb:64:in `block (2 levels) in class_from_fil
e'
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_cert.rb:51:in `each'
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_cert.rb:51:in `block in class_from_file'
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_certs.rb:16:in `block (2 levels) in class_from_fi
le'
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_certs.rb:15:in `each'
             /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_certs.rb:15:in `block in class_from_file'
             
             Resource Declaration:
             ---------------------
             # In /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_certs.rb
             
       16:     dns_acme_cert record['domain'] do
       17:       alt_names record['alt_names']
       18: 
       19:       contact new_resource.contact
       20:       acme_directory new_resource.acme_directory
       21:       acme_dns_api new_resource.acme_dns_api
       22:       cert_path "#{new_resource.cert_path}/#{record['domain']}.pem"
       23:       key_path "#{new_resource.key_path}/#{record['domain']}.key"
       24:     end
       25:   end
             
             Compiled Resource:
             ------------------
             # Declared in /tmp/kitchen/cache/cookbooks/osl-acme/resources/dns_acme_certs.rb:16:in `block (2 levels) in class_from_file'
             
             dns_acme_cert("test1.example.org") do
        action [:create]
        updated true
        updated_by_last_action true
        default_guard_interpreter :default
        declared_type :dns_acme_cert
        cookbook_name "acme_test"
        alt_names []
        contact "mailto:letsencrypt@osuosl.org"
        acme_directory "https://acme-staging-v02.api.letsencrypt.org/directory"
        acme_dns_api "http://192.168.10.1"
        cert_path "/etc/pki/tls/test1.example.org.pem"
        key_path "/etc/pki/tls/test1.example.org.key"
        domain "test1.example.org"
             end
             
             System Info:
             ------------
             chef_version=16.14.1
             platform=centos
             platform_version=7.9.2009
             ruby=ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux]
             program_name=/opt/cinc/bin/cinc-solo
             executable=/opt/cinc/bin/cinc-solo
```

**Comment by andrewda** [comment]
@ramereth Hmm, looks like that Pebble issue is back on CentOS 7... It fixed itself for a bit on my end, but that seems to be the same error. If you try converging a second time, does it work?

**Comment by ramereth** [comment]
@andrewda nope, happens again on a second and even third run

**Comment by andrewda** [comment]
Looks like adding a `Requires=pebble.service` to the acme-dns service resolves this problem for CentOS 7 on my end! Let me know if there's a prettier way you can think of implementing this:

https://github.com/osuosl-cookbooks/osl-acme/blob/53b99d3cc275002e699c26a2c5987237cf17ef7d/recipes/acme_dns_server.rb#L68-L90

File: recipes/acme_dns_server.rb [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+  match 'string'
+  extra_options '--dport 80 --string "POST /register" --algo kmp'
+  jump 'DROP'
+end
+
+unit_requires = if node['acme']['dir'] == 'https://127.0.0.1:14000/dir'
+                  'Requires=pebble.service'
+                else
+                  ''
+                end
```
Comment (detjensrobert) [inline recipes/acme_dns_server.rb:72]: No need for the else clause here:

```suggestion
unit_requires = 'Requires=pebble.service' if node['acme']['dir'] == 'https://127.0.0.1:14000/dir'
```

Might be good to add a comment for why this is needed.

Comment (detjensrobert) [reply recipes/acme_dns_server.rb:72]: The other way you could do this is with a resource drop-in in the test recipe, but this is simpler.

Comment (andrewda) [reply recipes/acme_dns_server.rb:72]: Good idea!

**Review by ramereth — APPROVED** [review]
Let's do a major bump for this.

**Comment by andrewda** [comment]
!bump major ~

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a major-level version bump to v4.0.0.  Have a nice day!
